openedx/features/qverse_features/registration/signals.py

```diff
@@ -1,20 +1,182 @@
+    """
+    Creates/Updates users and their profiles.
+
+    Reads data from given csv file having user details. Then creates/updates
+    edx user, edx user profile and qverse user profile accordingly. It also
+    writes the status of creation/update on the given file by adding an
+    extra column named 'status' in it. Also handles the exceptions raised by
+    the functions that it calls.
+
+    Arguments:
+        sender (ModelBase): responsible to initiate the signal
+        instance (BulkUserRegistration): newly created/updated object of BulkUserRegistration
+        created (bool): Is created/updated?
+        kwargs (dict): Other info
+    """
+    csv_file = None
+    try:
+        csv_file = open(instance.admission_file.path, 'r')
+    except IOError as error:
+        LOGGER.exception('({}) --- {}'.format(error.filename, error.strerror))
+        return
+    dict_reader = DictReader(csv_file, delimiter=',')
+    # The header of uploaded csv might be in uppercase, lowercase or in camel case
+    # so to make our code independent to case just turn all the keys of reader into lower case.
+    # Also in python standard library for csv there is an option named skipinitialwhitespaces
+    # which is only applied for leading whitespaces and there is no option for trailing
+    # whitespaces. So, we will have to handle that case ourselves
+    reader = (dict((k.strip().lower(), v.strip() if v else v) for k, v in row.items()) for row in dict_reader)
+    output_file_rows = []
+    try:
+        for row in reader:
+            try:
+                output_file_rows.append(row)
+                output_file_rows[-1].update({'status': ''})
+                _validate_single_csv_row(row)
+                # Creating/Updating edX User, edX User Profile and QVerse User Profile
+                # will be an atomic operation. If one operation fails the previous
+                # successfull operations will be reverted and no changes will be applied.
+                with transaction.atomic():
+                    edx_user = _create_or_update_edx_user(row)
+                    _create_or_update_edx_user_profile(edx_user)
+                    _create_or_update_qverse_user_profile(edx_user, row, output_file_rows)
+            except (IntegrityError, ValueError):
+                LOGGER.exception('Error while creating/updating ({}) user and its profiles.'.format(row.get('regno')))
+                output_file_rows[-1].update({'status': 'Failed'})
+            except Department.DoesNotExist:
+                LOGGER.exception('The Department with id {} does not exist.'.format(row.get('departmentid')))
+                output_file_rows[-1].update({'status': 'Failed'})
+    except Error as err:
+        LOGGER.exception('Error while traversing {} file content at line number {} with following error {}.'
+                         .format(instance.admission_file.path, reader.line_num, err))
+    csv_file.close()
```

Comment:
you can use simply output_row until here to avoid negative indexing and append here in `output_file_rows `

Reply:
done.